src/pages/Document.tsx

```diff
@@ -20,15 +31,70 @@ export default function Document() {
   /* Dotting */
   const ref = useRef<DottingRef>(null);
   const { dataArray } = useData(ref);
-  const { colorPixels, setData } = useDotting(ref);
+  const { colorPixels, setData, undo, redo, clear } = useDotting(ref);
   const { addStrokeEndListener, removeStrokeEndListener } = useHandlers(ref);
 
   /* Get data from hook */
   const { isGridFixed, isGridVisible, isPanZoomEnable } = useDottingContext();
   const { doc, client, isMultiplayerReady } = usePeer({ docId, dataArray, setData });
 
+  /* Manage History */
+  const [undoHistory, setUndoHistory] = useState<ColorChangeItem[][]>([]);
+  const [redoHistory, setRedoHistory] = useState<ColorChangeItem[][]>([]);
+
+  /* Manage Peer */
+  const { currentClient, peersExceptCurrentClient, syncPeers } = useDocumentContext();
+
+  /* Undo */
+  const undoData = useCallback(() => {
+    if (!undoHistory) return;
+    undo();
+    doc?.update((root) => {
+      const history = undoHistory.pop();
+      if (!history) return;
+
+      setRedoHistory([...redoHistory.map((items) => [...items]), history]);
+      history?.forEach((item) => {
+        const { color, rowIndex, columnIndex, previousColor } = item;
+        root.data[rowIndex][columnIndex].color = previousColor;
+      });
+    });
+  }, [dataArray]);
```

Comment:
It seems that you are using a custom undo, redo for only coloring data. I think it would be better if we updated the `DataChangeListener` in `dotting`. From what I know, currently `dotting` only emits the whole data without notifying the modified parts. If we updated the `DataChangeListener` to emit the modified data, I think we could solve this problem.

Reply:
The problem of this approach is that `undo` and `redo` will only be applicable for coloring data. Grid changes will not be appliable.

Reply:
@hunkim98 
Oh, I see. If It will be able to set the grid changes later (e.g. setIndices), I'll fix it together!